mlflow/store/ftp_artifact_repo.py

```diff
@@ -62,89 +62,67 @@ def _size(self, full_file_path):
-                        ftp.retrbinary("RETR "+ftp_file, f)
+    def get_path_module(self):
+        import posixpath
+        return posixpath
 
     def log_artifact(self, local_file, artifact_path=None):
         with self.get_ftp_client() as ftp:
-            artifact_dir = os.path.join(self.path, artifact_path) \
+            artifact_dir = self.get_path_module().join(self.path, artifact_path) \
                 if artifact_path else self.path
             self._mkdir(artifact_dir)
             with open(local_file, 'rb') as f:
                 ftp.cwd(artifact_dir)
-                ftp.storbinary('STOR ' + os.path.basename(local_file), f)
+                ftp.storbinary('STOR ' + self.get_path_module().basename(local_file), f)
 
     def log_artifacts(self, local_dir, artifact_path=None):
-        dest_path = os.path.join(self.path, artifact_path) \
+        dest_path = self.get_path_module().join(self.path, artifact_path) \
             if artifact_path else self.path
 
-        dest_path = build_path(dest_path, os.path.split(local_dir)[1])
-        dest_path_re = os.path.split(local_dir)[1]
+        dest_path = self.get_path_module().join(
+            dest_path, self.get_path_module().split(local_dir)[1])
+        dest_path_re = self.get_path_module().split(local_dir)[1]
         if artifact_path:
-            dest_path_re = build_path(artifact_path, os.path.split(local_dir)[1])
+            dest_path_re = self.get_path_module().join(
+                artifact_path, self.get_path_module().split(local_dir)[1])
 
-        local_dir = os.path.abspath(local_dir)
+        local_dir = self.get_path_module().abspath(local_dir)
         for (root, _, filenames) in os.walk(local_dir):
             upload_path = dest_path
             if root != local_dir:
-                rel_path = get_relative_path(local_dir, root)
-                upload_path = build_path(dest_path_re, rel_path)
+                rel_path = self.get_path_module().relpath(root, local_dir)
+                upload_path = self.get_path_module().join(dest_path_re, rel_path)
             if not filenames:
-                self._mkdir(build_path(self.path, upload_path))
+                self._mkdir(self.get_path_module().join(self.path, upload_path))
             for f in filenames:
-                if os.path.isfile(build_path(root, f)):
-                    self.log_artifact(build_path(root, f), upload_path)
+                if self.get_path_module().isfile(self.get_path_module().join(root, f)):
+                    self.log_artifact(self.get_path_module().join(root, f), upload_path)
 
     def list_artifacts(self, path=None):
         with self.get_ftp_client() as ftp:
             artifact_dir = self.path
-            list_dir = os.path.join(artifact_dir, path) if path else artifact_dir
+            list_dir = self.get_path_module().join(artifact_dir, path) if path else artifact_dir
+
+            print(list_dir, self._is_dir(list_dir))
+            if not self._is_dir(list_dir):
```

Comment:
Is this a correct behavior change, for list_artifacts to return an empty list as opposed to the object itself when called on a non-directory object?

Reply:
Yep! The existing documentation for `ArtifactRepository.list_artifacts` defines these semantics: https://github.com/mlflow/mlflow/blob/c8a2c489940436cdf73d09726876206c6f17c670/mlflow/store/artifact_repo.py#L46-L47